### PR TITLE
[FEATURE] Introduces SetBasedColumnMapExpectation w/ supporting templates & doc

### PR DESCRIPTION
**Pull request by abegong**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Introduces `SetBasedColumnMapExpectation` 
- Introduces template & example for above
- Introduces doc supporting above
- Updates sidebar

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:

### Definition of Done
Please delete options that are not relevant.

Thank you for submitting!

### CODE REVIEW

**Comment by abegong** [comment]
Loom: [SetBasedColumnMapExpectations - 21 March 2022](https://www.loom.com/share/94e60eafd77844cba0831438b2f81b9f)

Todo:
* Implement sqlalchemy and spark
* Make sure that this approach to `validate_configuration` works.
* Implement other renders
* Flesh out the How-to Guide
* Proofread
* Friction log

To do separately:
* Remove `validate_configuration` and corresponding `import`s from `RegexBasedColumnMapExpectation` template and docs.

**Comment by austiezr** [comment]
@abegong I've taken a pass at the sqlalchemy & spark implementations, renderers, and how-to guide. Also updated the regex templates & guide (#4506)

File: docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+2. They help users understand the logic of your Expectation by providing tidy examples of paired input and output. If you contribute your Expectation to open source, these examples will appear in the Gallery.
+
+Your examples will look something like this:
+
+```python file=../../../../tests/integration/docusaurus/expectations/creating_custom_expectations/expect_column_values_to_be_in_solfege_scale_set.py#L37-L117
+```
+
+Here's a quick overview of how to create test cases to populate `examples`. The overall structure is a list of dictionaries. Each dictionary has two keys:
+
+* `data`: defines the input data of the example as a table/data frame. In this example the table has columns named `lowercase_solfege_scale`, `uppdecase_solfege_scale`, and `mixed`. All of these columns have 8 rows. (Note: if you define multiple columns, make sure that they have the same number of rows.)
```
Comment (talagluck) [inline docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md]:
```suggestion
* `data`: defines the input data of the example as a table/data frame. In this example the table has columns named `lowercase_solfege_scale`, `uppercase_solfege_scale`, and `mixed`. All of these columns have 8 rows. (Note: if you define multiple columns, make sure that they have the same number of rows.)
```

File: docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+```python file=../../../../examples/expectations/set_based_column_map_expectation_template.py#L18-L21
+```
+
+with something like this:
+
+```python file=../../../../tests/integration/docusaurus/expectations/creating_custom_expectations/expect_column_values_to_be_in_solfege_scale_set.py#L10-L34
+```
+
+For more detail when rendering your Custom Expectation, you can optionally specify the plural form of a Semantic Type you're validating.
```
Comment (talagluck) [inline docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md]: Is there a reason we need to specify both the `set_snake_name` and `set_camel_name`? Couldn't we translate from one to the other [e.g. using this](https://github.com/great-expectations/great_expectations/blob/29c6362ad287d1420633560246cb11d163872bbf/great_expectations/util.py#L128)?

Could you also provide some more explanation about the plural form? How does this align with the `set_semantic_name`?

**Review by talagluck — COMMENTED** [review]
Looks great overall! Just a couple of questions.

File: docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md (hunk 2 above)
Comment (austiezr) [reply docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md]: To point one: I don't see why not! 🙇 

To point two: Mis-copy on my part. This should be updated to reference set semantic name.

Comment (abegong) [reply docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md]: ❤️

Comment (austiezr) [reply docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations.md]: Updated re: above

**Review by talagluck — APPROVED** [review]
LGTM - this is awesome!

**Review by donaldheppner — APPROVED** [review]

**Pull request merged**